torch/csrc/jit/codegen/cuda/expr_simplifier.cpp

```diff
@@ -0,0 +1,438 @@
+#include <torch/csrc/jit/codegen/cuda/expr_simplifier.h>
+
+#include <torch/csrc/jit/codegen/cuda/ir_all_nodes.h>
+#include <torch/csrc/jit/codegen/cuda/ir_builder.h>
+#include <torch/csrc/jit/codegen/cuda/ir_cloner.h>
+
+#include <unordered_set>
+#include <vector>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+
+namespace {
+
+bool hasSimilarType(DataType t1, DataType t2) {
+  if (t1 == t2) {
+    return true;
+  }
+  if (isIntegralType(t1) && isIntegralType(t2)) {
+    return true;
+  }
+  if (isFloatingPointType(t1) && isFloatingPointType(t2)) {
+    return true;
+  }
+  if (isComplexType(t1) && isComplexType(t2)) {
+    return true;
+  }
+  return false;
+}
+
+bool isMagicZeroProtectedVariable(Val* value) {
```

Comment:
There's `isProtectedWithMagicZero` in lowre_magic_zero.h

Reply:
fixed in https://github.com/csarofeen/pytorch/pull/2257